File: leetcode/Algorithms/378.KthSmallestElementInASortedMatrix/Solution.py
class Solution(object):
    def kthSmallest(self, matrix, k):
        """
        :type matrix: List[List[int]]
        :type k: int
        :rtype: int
        """
        low, high = matrix[0][0], matrix[len(matrix) - 1][len(matrix[0]) - 1] + 1
        while low < high:
            mid = (high + low) / 2
            if self.binary_search(matrix, mid) < k:
                low = mid + 1
            else:
                high = mid
        return low

    def binary_search(self, matrix, mid):
        i, j, result = len(matrix) - 1, 0, 0
        while i >= 0 and j <= len(matrix) - 1:
            if matrix[i][j] > mid:
                i -= 1
            else:
                j += 1
                result += (i + 1)
        return result

File: leetcode/Algorithms/378.KthSmallestElementInASortedMatrix/Solution.java
import java.util.PriorityQueue;

public class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        PriorityQueue<Tuple> queue = new PriorityQueue<>();
        for (int j = 0; j < matrix.length; j++) queue.offer(new Tuple(0, j, matrix[0][j]));

        while (--k != 0) {
            Tuple entry = queue.poll();
            if (entry.x < matrix.length - 1) {
                queue.offer(new Tuple(entry.x + 1, entry.y, matrix[entry.x + 1][entry.y]));
            }
        }

        return queue.poll().val;
    }
}

class Tuple implements Comparable<Tuple> {
    int x, y, val;

    Tuple(int x, int y, int val) {
        this.x = x;
        this.y = y;
        this.val = val;
    }

    @Override
    public int compareTo(Tuple that) {
        return this.val - that.val;
    }
}
